lib/qr-bills/qr-generator.rb

```diff
@@ -1,6 +1,86 @@
+# frozen_string_literal: true
+
 require 'rqrcode'
 
 module QRGenerator
+  def self.create(bill_params, format: 'png')
+    case format
+    when 'qrcode_png', 'png'
+      build_png(bill_params)
+    when 'svg'
+      build_svg(bill_params)
+    else
+      raise ArgumentError, "#{QRExceptions::NOT_SUPPORTED}: #{format} is not yet supported"
+    end
+  end
+
+  def self.build_png(bill_params)
+    payload = build_payload(bill_params)
+    qrcode = RQRCode::QRCode.new(payload, level: 'm')
+
+    png = qrcode.as_png(
+      bit_depth: 1,
+      border_modules: 0,
+      color_mode: ChunkyPNG::COLOR_GRAYSCALE,
+      color: 'black',
+      file: nil,
+      fill: 'white',
+      module_px_size: 10,
+      resize_exactly_to: false,
+      resize_gte_to: false,
+      size: 1024
+    )
+
+    swiss_cross = ChunkyPNG::Image.from_file(File.expand_path("#{File.dirname(__FILE__)}/../../web/assets/images/swiss_cross.png"))
+    final_qr = png.compose!(swiss_cross, (png.width - swiss_cross.width) / 2, (png.height - swiss_cross.height) / 2)
+    final_qr
+  end
+
+  def self.build_svg(bill_params)
+    payload = build_payload(bill_params)
+    qrcode = RQRCode::QRCode.new(payload, level: 'm')
+
+    # Generate qr code
+    svg_code = qrcode.as_svg(
+      standalone: true, # return a svg wrapper
+      viewbox: true, # return a viewBox so we can extract it
+      module_size: 10,
+      use_path: true
+    )
+
+    # Parse viewBox and calculate relative swiss cross size / center
+    viewbox = svg_code.match(/viewBox="([\s\d]*)"/)[1]
+    viewbox_size = viewbox.split(' ')[2].to_f
+    swiss_cross_size = viewbox_size / (46.0 / 7.0)
+    swiss_cross_center = (viewbox_size / 2.0) - (swiss_cross_size / 2.0)
+
+    # Parse swiss cross and update size and coordinates
+    svg_swiss_cross = File.open(File.expand_path("#{File.dirname(__FILE__)}/../../web/assets/images/swiss_cross.svg")).read
+    svg_swiss_cross.gsub!('SWISS_CROSS_SIZE', swiss_cross_size.to_s)
+    svg_swiss_cross.gsub!('SWISS_CROSS_CENTER', swiss_cross_center.to_s)
+
+    # Assemble svg
+    svg_code.gsub!(/<path/, '<g>\0')
+    svg_code.gsub!(/<\/svg>/, '</g>\0')
+    svg_code.gsub!(/<\/g>/, "#{svg_swiss_cross}</g>")
+
+    # Optimize SVG for being URI-escaped
```

Comment:
It's because of encoding within <img src="">. Single quotes do not need encoding within double quotes which gives a slightly better performance. The SVG comments are not needed as well. For the new lines / whitespace removal I think we can address this directly within the swiss_cross.svg file as rqrcode will already produce a slim svg string.

Reply:
ok thx!